File: packages/observability-py/src/flownote_observability/schema.py
"""ログレコードのスキーマ定義。

[ログ規約](../../../docs/observability/logging-spec.md) §2 のスキーマを Pydantic で固定する。
本モデルは「構造化ログが規約に準拠するか」をテストで検証するための唯一の実装である。
"""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field

# OTel リソース属性のドット区切りキー(トップレベル)。
SERVICE_NAME_KEY = "service.name"
SERVICE_VERSION_KEY = "service.version"
# OTel Stable 化で `deployment.environment` は `deployment.environment.name` に rename された。
DEPLOYMENT_ENVIRONMENT_KEY = "deployment.environment.name"
# ログスキーマのバージョン。規約進化時に ETL/パース側が世代を判別できるよう必須で付与する。
LOG_SCHEMA_VERSION_KEY = "flownote.log.schema_version"
LOG_SCHEMA_VERSION = "1"


class LogRecord(BaseModel):
    """規約準拠の構造化ログ1件。

    Pydantic の別名(alias)で OTel のドット区切りキー(``service.name`` 等)を表現する。
    ``populate_by_name=True`` により Python 名でも別名でも生成できる。

    Attributes:
        timestamp: RFC3339(ナノ秒)・UTC の発生時刻。
        severity_text: 重大度ラベル (``INFO`` 等)。
        severity_number: OTel SeverityNumber。
        body: 低カーディナリティのイベント名/メッセージ。
        log_schema_version: ログスキーマ世代(別名 ``flownote.log.schema_version``)。
        service_name: 発生元サービス(別名 ``service.name``)。
        service_version: バージョン(別名 ``service.version``)。
        deployment_environment: 環境(別名 ``deployment.environment.name``)。
        trace_id: 相関トレースID。span 外なら ``None``。
        span_id: 相関 span ID。
        attributes: 構造化属性。
    """

    model_config = ConfigDict(populate_by_name=True, extra="forbid")

    timestamp: str
    severity_text: str
    severity_number: int
    body: str
    log_schema_version: str = Field(alias="flownote.log.schema_version")
    # alias は mypy/pydantic プラグインの要請でリテラル文字列を用いる
    # (値は上の *_KEY 定数と一致させること)。
    service_name: str = Field(alias="service.name")
    service_version: str = Field(alias="service.version")
    deployment_environment: str = Field(alias="deployment.environment.name")
    trace_id: str | None = None
    span_id: str | None = None
    # 属性は任意の JSON 互換値を取りうる異種バッグのため object 型とする
    # (``Any`` は使わない方針。値の健全性はマスキングと出力時の JSON 化で担保)。
    attributes: dict[str, object] = Field(default_factory=dict)


# トップレベルに置かれる予約キー(これら以外は ``attributes`` へ集約される)。
RESERVED_TOP_LEVEL_KEYS: frozenset[str] = frozenset(
    {
        "timestamp",
        "severity_text",
        "severity_number",
        "body",
        LOG_SCHEMA_VERSION_KEY,
        SERVICE_NAME_KEY,
        SERVICE_VERSION_KEY,
        DEPLOYMENT_ENVIRONMENT_KEY,
        "trace_id",
        "span_id",
        "attributes",
    }
)

File: packages/observability-web/src/schema.ts
/**
 * クライアントログのスキーマ。
 *
 * バックエンドの {@link ../../../docs/observability/logging-spec.md} §2 と整合する形に整形する。
 */

import { ATTR } from "./semconv";
import type { SeverityText } from "./severity";
import { SEVERITY_NUMBER } from "./severity";

/** 構造化クライアントログ1件。 */
export interface ClientLogRecord {
  /** RFC3339(ミリ秒)・UTC の発生時刻。 */
  timestamp: string;
  /** 重大度ラベル。 */
  severity_text: SeverityText;
  /** OTel SeverityNumber。 */
  severity_number: number;
  /** 低カーディナリティのイベント名。 */
  body: string;
  /** 発生元サービス(常に flownote-web)。 */
  "service.name": string;
  /** バージョン。 */
  "service.version": string;
  /** 環境(OTel Stable: `deployment.environment.name`。BE/SSOT と一致)。 */
  "deployment.environment.name": string;
  /** 相関トレースID。 */
  trace_id: string | null;
  /** 相関スパンID。 */
  span_id: string | null;
  /** 構造化属性。 */
  attributes: Record<string, unknown>;
}

/** ログ整形に必要なリソース情報。 */
export interface ResourceInfo {
  serviceName: string;
  serviceVersion: string;
  environment: string;
}

/**
 * スキーマ準拠のログレコードを組み立てる。
 *
 * @param params - レコード生成のパラメータ。
 * @param params.severity - 重大度ラベル。
 * @param params.body - イベント名。
 * @param params.attributes - 構造化属性。
 * @param params.resource - リソース情報。
 * @param params.trace - 相関トレース文脈(無ければ null)。
 * @param params.nowIso - 生成時刻の ISO 文字列(テスト容易性のため注入可能)。
 * @returns 整形済みの {@link ClientLogRecord}。
 */
export function buildRecord(params: {
  severity: SeverityText;
  body: string;
  attributes: Record<string, unknown>;
  resource: ResourceInfo;
  trace: { traceId: string; spanId: string } | null;
  nowIso: string;
}): ClientLogRecord {
  return {
    timestamp: params.nowIso,
    severity_text: params.severity,
    severity_number: SEVERITY_NUMBER[params.severity],
    body: params.body,
    [ATTR.SERVICE_NAME]: params.resource.serviceName,
    [ATTR.SERVICE_VERSION]: params.resource.serviceVersion,
    [ATTR.DEPLOYMENT_ENVIRONMENT]: params.resource.environment,
    trace_id: params.trace?.traceId ?? null,
    span_id: params.trace?.spanId ?? null,
    attributes: params.attributes,
  };
}
